File: py/koladata/functor/signature_utils_test.py
# Copyright 2024 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Tests for signature_utils."""

import inspect

from absl.testing import absltest
from koladata.functions import functions as fns
from koladata.functor import signature_utils
from koladata.operators import kde_operators as _
from koladata.types import data_item
from koladata.types import data_slice
from koladata.types import schema_constants


class SignatureUtilsTest(absltest.TestCase):

  def test_parameter(self):
    p = signature_utils.parameter(
        'x', signature_utils.ParameterKind.POSITIONAL_ONLY
    )
    self.assertEqual(p.get_schema(), schema_constants.OBJECT)
    self.assertEqual(p.name, 'x')
    self.assertEqual(p.kind, signature_utils.ParameterKind.POSITIONAL_ONLY)
    self.assertEqual(p.default_value, signature_utils.NO_DEFAULT_VALUE)

  def test_parameter_with_default(self):
    p = signature_utils.parameter(
        'x', signature_utils.ParameterKind.POSITIONAL_ONLY, 57
    )
    self.assertEqual(p.default_value, 57)
    self.assertIsInstance(p.default_value, data_item.DataItem)

  def test_parameter_with_default_slice(self):
    with self.assertRaisesRegex(
        ValueError, 'only DataItems can be used as default values'
    ):
      _ = signature_utils.parameter(
          'x',
          signature_utils.ParameterKind.POSITIONAL_ONLY,
          data_slice.DataSlice.from_vals([1, 2, 3]),
      )

  def test_parameter_with_default_none(self):
    p = signature_utils.parameter(
        'x', signature_utils.ParameterKind.POSITIONAL_ONLY, None
    )
    self.assertEqual(p.default_value.get_present_count(), 0)
    self.assertEqual(p.default_value.get_schema(), schema_constants.NONE)

  def test_parameter_adopts_default(self):
    v = fns.new(foo=57)
    p = signature_utils.parameter(
        'x', signature_utils.ParameterKind.POSITIONAL_ONLY, v
    )
    self.assertEqual(p.default_value, v)
    self.assertEqual(p.default_value.foo, 57)

  def test_signature(self):
    p1 = signature_utils.parameter(
        'x', signature_utils.ParameterKind.POSITIONAL_ONLY
    )
    p2 = signature_utils.parameter(
        'y', signature_utils.ParameterKind.POSITIONAL_OR_KEYWORD
    )
    s = signature_utils.signature([p1, p2])
    self.assertEqual(s.get_schema(), schema_constants.OBJECT)
    self.assertEqual(s.parameters[:].name.to_py(), ['x', 'y'])
    self.assertEqual(s.parameters[:].to_py(), [p1, p2])
    self.assertFalse(s.db.is_mutable())

  def test_from_py_signature(self):
    entity = fns.new(foo=57)
    py_sig = inspect.Signature([
        inspect.Parameter('p1', inspect.Parameter.POSITIONAL_ONLY, default=1),
        inspect.Parameter(
            'p2', inspect.Parameter.POSITIONAL_OR_KEYWORD, default=entity
        ),
        inspect.Parameter('p3', inspect.Parameter.VAR_POSITIONAL),
        inspect.Parameter('p4', inspect.Parameter.KEYWORD_ONLY),
        inspect.Parameter('p5', inspect.Parameter.VAR_KEYWORD),
    ])
    sig = signature_utils.from_py_signature(py_sig)
    self.assertEqual(
        sig.parameters[:].name.to_py(),
        ['p1', 'p2', 'p3', 'p4', 'p5'],
    )
    self.assertEqual(
        sig.parameters[:].kind.to_py(),
        [
            signature_utils.ParameterKind.POSITIONAL_ONLY,
            signature_utils.ParameterKind.POSITIONAL_OR_KEYWORD,
            signature_utils.ParameterKind.VAR_POSITIONAL,
            signature_utils.ParameterKind.KEYWORD_ONLY,
            signature_utils.ParameterKind.VAR_KEYWORD,
        ],
    )
    self.assertEqual(sig.parameters[0].default_value, 1)
    self.assertEqual(sig.parameters[1].default_value, entity)
    self.assertEqual(
        sig.parameters[2].default_value, signature_utils.NO_DEFAULT_VALUE
    )
    self.assertEqual(
        sig.parameters[3].default_value, signature_utils.NO_DEFAULT_VALUE
    )
    self.assertEqual(
        sig.parameters[4].default_value, signature_utils.NO_DEFAULT_VALUE
    )

  def test_from_py_signature_slice_as_default(self):
    py_sig = inspect.Signature([
        inspect.Parameter(
            'p1',
            inspect.Parameter.POSITIONAL_ONLY,
            default=data_slice.DataSlice.from_vals([1, 2, 3]),
        ),
    ])
    with self.assertRaisesRegex(
        ValueError, 'only DataItems can be used as default values'
    ):
      _ = signature_utils.from_py_signature(py_sig)

  def test_arg_kwargs_signature(self):
    sig = signature_utils.ARGS_KWARGS_SIGNATURE
    self.assertEqual(
        sig.parameters[:].name.to_py(),
        ['args', 'kwargs'],
    )
    self.assertEqual(
        sig.parameters[:].kind.to_py(),
        [
            signature_utils.ParameterKind.VAR_POSITIONAL,
            signature_utils.ParameterKind.VAR_KEYWORD,
        ],
    )


if __name__ == '__main__':
  absltest.main()

File: koladata/functor/signature_utils_test.cc
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
#include "koladata/functor/signature_utils.h"

#include <optional>
#include <vector>

#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "absl/status/status.h"
#include "absl/status/status_matchers.h"
#include "arolla/qtype/tuple_qtype.h"
#include "arolla/qtype/typed_ref.h"
#include "koladata/data_slice.h"
#include "koladata/functor/signature.h"
#include "koladata/internal/dtype.h"
#include "koladata/test_utils.h"
#include "koladata/testing/matchers.h"

namespace koladata::functor {

namespace {

using ::absl_testing::IsOkAndHolds;
using ::absl_testing::StatusIs;
using ::koladata::testing::IsEquivalentTo;
using ::testing::HasSubstr;

TEST(BindArgumentsTest, Basic) {
  Signature::Parameter p1 = {
      .name = "a",
      .kind = Signature::Parameter::Kind::kPositionalOnly,
  };
  Signature::Parameter p2 = {
      .name = "b",
      .kind = Signature::Parameter::Kind::kPositionalOrKeyword,
      .default_value = test::DataItem(1),
  };
  Signature::Parameter p3 = {
      .name = "c",
      .kind = Signature::Parameter::Kind::kVarPositional,
  };
  Signature::Parameter p4 = {
      .name = "d",
      .kind = Signature::Parameter::Kind::kKeywordOnly,
      .default_value = test::DataItem(std::nullopt, schema::kInt32),
  };
  Signature::Parameter p5 = {
      .name = "e",
      .kind = Signature::Parameter::Kind::kVarKeyword,
  };
  ASSERT_OK_AND_ASSIGN(auto signature, Signature::Create({p1, p2, p3, p4, p5}));
  std::vector<DataSlice> slices = {test::DataItem(0), test::DataItem(1),
                                   test::DataItem(2), test::DataItem(3),
                                   test::DataItem(4), test::DataItem(5)};
  ASSERT_OK_AND_ASSIGN(auto bound_arguments,
                       BindArguments(signature,
                                     {arolla::TypedRef::FromValue(slices[0]),
                                      arolla::TypedRef::FromValue(slices[1]),
                                      arolla::TypedRef::FromValue(slices[2]),
                                      arolla::TypedRef::FromValue(slices[3]),
                                      arolla::TypedRef::FromValue(slices[4]),
                                      arolla::TypedRef::FromValue(slices[5])},
                                     {"f", "g"}));
  auto expected_var_args =
      arolla::MakeTuple({arolla::TypedRef::FromValue(slices[2]),
                         arolla::TypedRef::FromValue(slices[3])});
  ASSERT_OK_AND_ASSIGN(
      auto expected_var_kwargs,
      arolla::MakeNamedTuple({"f", "g"},
                             {arolla::TypedRef::FromValue(slices[4]),
                              arolla::TypedRef::FromValue(slices[5])}));
  ASSERT_EQ(bound_arguments.size(), 5);
  EXPECT_THAT(bound_arguments[0].As<DataSlice>(),
              IsOkAndHolds(IsEquivalentTo(slices[0])));
  EXPECT_THAT(bound_arguments[1].As<DataSlice>(),
              IsOkAndHolds(IsEquivalentTo(slices[1])));
  EXPECT_THAT(bound_arguments[2].GetFingerprint(),
              expected_var_args.GetFingerprint());
  EXPECT_THAT(bound_arguments[3].As<DataSlice>(),
              IsOkAndHolds(IsEquivalentTo(p4.default_value.value())));
  EXPECT_THAT(bound_arguments[4].GetFingerprint(),
              expected_var_kwargs.GetFingerprint());
}

TEST(BindArgumentsTest, NotValidIdentifiers) {
  Signature::Parameter p1 = {
      .name = " ",
      .kind = Signature::Parameter::Kind::kPositionalOnly,
  };
  Signature::Parameter p2 = {
      .name = "?",
      .kind = Signature::Parameter::Kind::kPositionalOrKeyword,
  };
  Signature::Parameter p3 = {
      .name = "-",
      .kind = Signature::Parameter::Kind::kVarPositional,
  };
  Signature::Parameter p4 = {
      .name = "!",
      .kind = Signature::Parameter::Kind::kKeywordOnly,
  };
  Signature::Parameter p5 = {
      .name = "=",
      .kind = Signature::Parameter::Kind::kVarKeyword,
  };
  ASSERT_OK_AND_ASSIGN(auto signature, Signature::Create({p1, p2, p3, p4, p5}));
  std::vector<DataSlice> slices = {test::DataItem(0), test::DataItem(1),
                                   test::DataItem(2), test::DataItem(3),
                                   test::DataItem(4), test::DataItem(5)};
  ASSERT_OK_AND_ASSIGN(auto bound_arguments,
                       BindArguments(signature,
                                     {arolla::TypedRef::FromValue(slices[0]),
                                      arolla::TypedRef::FromValue(slices[1]),
                                      arolla::TypedRef::FromValue(slices[2]),
                                      arolla::TypedRef::FromValue(slices[3]),
                                      arolla::TypedRef::FromValue(slices[4]),
                                      arolla::TypedRef::FromValue(slices[5])},
                                     {"", "!"}));
  auto expected_var_args =
      arolla::MakeTuple({arolla::TypedRef::FromValue(slices[2]),
                         arolla::TypedRef::FromValue(slices[3])});
  ASSERT_OK_AND_ASSIGN(
      auto expected_var_kwargs,
      arolla::MakeNamedTuple({""}, {arolla::TypedRef::FromValue(slices[4])}));
  ASSERT_EQ(bound_arguments.size(), 5);
  EXPECT_THAT(bound_arguments[0].As<DataSlice>(),
              IsOkAndHolds(IsEquivalentTo(slices[0])));
  EXPECT_THAT(bound_arguments[1].As<DataSlice>(),
              IsOkAndHolds(IsEquivalentTo(slices[1])));
  EXPECT_EQ(bound_arguments[2].GetFingerprint(),
            expected_var_args.GetFingerprint());
  EXPECT_THAT(bound_arguments[3].As<DataSlice>(),
              IsOkAndHolds(IsEquivalentTo(slices[5])));
  EXPECT_EQ(bound_arguments[4].GetFingerprint(),
            expected_var_kwargs.GetFingerprint());
}

TEST(BindArgumentsTest, PositionalOnlyAndKeywordSameName) {
  Signature::Parameter p1 = {
      .name = "foo",
      .kind = Signature::Parameter::Kind::kPositionalOnly,
      .default_value = test::DataItem(57),
  };
  Signature::Parameter p2 = {
      .name = "kwargs",
      .kind = Signature::Parameter::Kind::kVarKeyword,
  };
  ASSERT_OK_AND_ASSIGN(auto signature, Signature::Create({p1, p2}));
  auto input_slice = test::DataItem(43);
  ASSERT_OK_AND_ASSIGN(
      auto bound_arguments,
      BindArguments(signature, {arolla::TypedRef::FromValue(input_slice)},
                    {"foo"}));
  ASSERT_EQ(bound_arguments.size(), 2);
  EXPECT_THAT(bound_arguments[0].As<DataSlice>(),
              IsOkAndHolds(IsEquivalentTo(p1.default_value.value())));
  ASSERT_OK_AND_ASSIGN(
      auto expected_kwargs,
      arolla::MakeNamedTuple({"foo"},
                             {arolla::TypedRef::FromValue(input_slice)}));
  EXPECT_EQ(bound_arguments[1].GetFingerprint(),
            expected_kwargs.GetFingerprint());
}

TEST(BindArgumentsTest, PositionalOnlyAndKeywordSameNameErrorMessage) {
  Signature::Parameter p1 = {
      .name = "foo",
      .kind = Signature::Parameter::Kind::kPositionalOnly,
  };
  Signature::Parameter p2 = {
      .name = "kwargs",
      .kind = Signature::Parameter::Kind::kVarKeyword,
  };
  ASSERT_OK_AND_ASSIGN(auto signature, Signature::Create({p1, p2}));
  auto input_slice = test::DataItem(43);
  EXPECT_THAT(
      BindArguments(signature, {arolla::TypedRef::FromValue(input_slice)},
                    {"foo"}),
      StatusIs(absl::StatusCode::kInvalidArgument,
               "no value provided for positional only parameter [foo]"));
}

TEST(BindArgumentsTest, SpecifyingSameBothAsPositionalAndKeyword) {
  Signature::Parameter p1 = {
      .name = "foo",
      .kind = Signature::Parameter::Kind::kPositionalOrKeyword,
  };
  ASSERT_OK_AND_ASSIGN(auto signature, Signature::Create({p1}));
  auto input_slice = test::DataItem(43);
  EXPECT_THAT(BindArguments(signature,
                            {arolla::TypedRef::FromValue(input_slice),
                             arolla::TypedRef::FromValue(input_slice)},
                            {"foo"}),
              StatusIs(absl::StatusCode::kInvalidArgument,
                       "parameter [foo] specified twice"));
}

TEST(BindArgumentsTest, SpecifyingSameAsKeywordTwice) {
  Signature::Parameter p1 = {
      .name = "foo",
      .kind = Signature::Parameter::Kind::kPositionalOrKeyword,
  };
  ASSERT_OK_AND_ASSIGN(auto signature, Signature::Create({p1}));
  auto input_slice = test::DataItem(43);
  EXPECT_THAT(BindArguments(signature,
                            {arolla::TypedRef::FromValue(input_slice),
                             arolla::TypedRef::FromValue(input_slice)},
                            {"foo", "foo"}),
              StatusIs(absl::StatusCode::kInvalidArgument,
                       "parameter [foo] specified twice"));
}

TEST(BindArgumentsTest, SpecifyingSameUnknownAsKeywordTwice) {
  Signature::Parameter p1 = {
      .name = "foo",
      .kind = Signature::Parameter::Kind::kVarKeyword,
  };
  ASSERT_OK_AND_ASSIGN(auto signature, Signature::Create({p1}));
  auto input_slice = test::DataItem(43);
  EXPECT_THAT(BindArguments(signature,
                            {arolla::TypedRef::FromValue(input_slice),
                             arolla::TypedRef::FromValue(input_slice)},
                            {"foo", "foo"}),
              StatusIs(absl::StatusCode::kInvalidArgument,
                       // We can improve this error message if needed.
                       HasSubstr("field name foo is duplicated")));
}

TEST(BindArgumentsTest, TooFewPositional) {
  Signature::Parameter p1 = {
      .name = "foo",
      .kind = Signature::Parameter::Kind::kPositionalOnly,
  };
  ASSERT_OK_AND_ASSIGN(auto signature, Signature::Create({p1}));
  EXPECT_THAT(
      BindArguments(signature, {}, {}),
      StatusIs(absl::StatusCode::kInvalidArgument,
               "no value provided for positional only parameter [foo]"));
}

TEST(BindArgumentsTest, TooFewPositionalOkWithDefaultValue) {
  Signature::Parameter p1 = {
      .name = "foo",
      .kind = Signature::Parameter::Kind::kPositionalOnly,
      .default_value = test::DataItem(57),
  };
  ASSERT_OK_AND_ASSIGN(auto signature, Signature::Create({p1}));
  ASSERT_OK_AND_ASSIGN(auto bound_arguments, BindArguments(signature, {}, {}));
  EXPECT_THAT(bound_arguments[0].As<DataSlice>(),
              IsOkAndHolds(IsEquivalentTo(p1.default_value.value())));
}

TEST(BindArgumentsTest, TooManyPositional) {
  Signature::Parameter p1 = {
      .name = "foo",
      .kind = Signature::Parameter::Kind::kPositionalOnly,
  };
  ASSERT_OK_AND_ASSIGN(auto signature, Signature::Create({p1}));
  auto input_slice = test::DataItem(43);
  EXPECT_THAT(BindArguments(signature,
                            {arolla::TypedRef::FromValue(input_slice),
                             arolla::TypedRef::FromValue(input_slice),
                             arolla::TypedRef::FromValue(input_slice)},
                            {}),
              StatusIs(absl::StatusCode::kInvalidArgument,
                       "too many positional arguments passed (2 extra)"));
}

TEST(BindArgumentsTest, UnknownKeyword) {
  Signature::Parameter p1 = {
      .name = "foo",
      .kind = Signature::Parameter::Kind::kVarPositional,
  };
  ASSERT_OK_AND_ASSIGN(auto signature, Signature::Create({p1}));
  auto input_slice = test::DataItem(43);
  EXPECT_THAT(BindArguments(signature,
                            {arolla::TypedRef::FromValue(input_slice),
                             arolla::TypedRef::FromValue(input_slice)},
                            {"b", "a"}),
              StatusIs(absl::StatusCode::kInvalidArgument,
                       "unknown keyword arguments: [b, a]"));
}

TEST(BindArgumentsTest, InvalidArgsKwnamesSizes) {
  Signature::Parameter p1 = {
      .name = "foo",
      .kind = Signature::Parameter::Kind::kVarPositional,
  };
  ASSERT_OK_AND_ASSIGN(auto signature, Signature::Create({p1}));
  EXPECT_THAT(BindArguments(signature, {}, {"foo"}),
              StatusIs(absl::StatusCode::kInvalidArgument,
                       "args.size < kwnames.size()"));
}

}  // namespace

}  // namespace koladata::functor
